proc_smaps.go

```diff
@@ -0,0 +1,246 @@
+	Referenced uint64
+	// Amount of memory that does not belong to any file
+	Anonymous uint64
+	// Amount would-be-anonymous memory currently on swap
+	Swap uint64
+	// Process's proportional memory on swap
+	SwapPss uint64
+	// Page size used by the kernel to back the virtual memory area
+	KernelPageSize uint64
+	// Page size used by the processor MMU to back the virtual memory area
+	MMUPageSize uint64
+	// Kernel flags associated with the virtual memory area
+	VMFlags string
+}
+
+// SizeSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SizeSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Size
+	}
+
+	return
+}
+
+// RssSum returns the sum of Pss from all mappings
+func (s ProcSMaps) RssSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Rss
+	}
+
+	return
+}
+
+// PssSum returns the sum of Pss from all mappings
+func (s ProcSMaps) PssSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Pss
+	}
+
+	return
+}
+
+// SharedCleanSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SharedCleanSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.SharedClean
+	}
+
+	return
+}
+
+// SharedDirtySum returns the sum of Pss from all mappings
+func (s ProcSMaps) SharedDirtySum() (sum uint64) {
+	for _, x := range s {
+		sum += x.SharedDirty
+	}
+
+	return
+}
+
+// PrivateCleanSum returns the sum of Pss from all mappings
+func (s ProcSMaps) PrivateCleanSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.PrivateClean
+	}
+
+	return
+}
+
+// PrivateDirtySum returns the sum of Pss from all mappings
+func (s ProcSMaps) PrivateDirtySum() (sum uint64) {
+	for _, x := range s {
+		sum += x.PrivateDirty
+	}
+
+	return
+}
+
+// ReferencedSum returns the sum of Pss from all mappings
+func (s ProcSMaps) ReferencedSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Referenced
+	}
+
+	return
+}
+
+// AnonymousSum returns the sum of Pss from all mappings
+func (s ProcSMaps) AnonymousSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Anonymous
+	}
+
+	return
+}
+
+// SwapSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SwapSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Swap
+	}
+
+	return
+}
+
+// SwapPssSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SwapPssSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.SwapPss
+	}
+
+	return
+}
+
+// ProcSMaps reads from /proc/[pid]/maps to get the memory-mappings of the
+// process.
+func (p Proc) ProcSMaps() (ProcSMaps, error) {
+	file, err := os.Open(p.path("smaps"))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var currentMap *ProcSMap
+
+	smaps := []*ProcSMap{}
+	scan := bufio.NewScanner(file)
+
+	for scan.Scan() {
+		line := scan.Text()
+		// first line of a mapping start with an hexadecimal address in lower-case
+		// All other line start with a capitalized words.
+		if (line[0] >= '0' && line[0] <= '9') || (line[0] >= 'a' && line[0] <= 'f') {
```

Comment:
I think it would be more clear hear to do a simple regex match.

Reply:
I've kept is as an if condition for performance reason.
I've added a benchmark and pushed https://github.com/PierreF/procfs/commit/ff6dfede807f15a2d7ad605249ad41aa93184389 which use regexp, but it has notable performance impact:
```
$ git co 6481fbb # commit just before using regexp
$ for i in `seq 0 10`; do go test -run=none -bench=.* ./ ;done > bench.6481fbb
$ git co ff6dfed # commit using regexp
$ for i in `seq 0 10`; do go test -run=none -bench=.* ./ ;done > bench.ff6dfed
$ go get golang.org/x/perf/cmd/benchstat
$ benchstat bench.6481fbb bench.ff6dfed 
name         old time/op  new time/op  delta
ProcSmaps-8   200µs ± 8%   270µs ± 5%  +34.52%  (p=0.000 n=11+10)
```

I think this the less clear hand-made condition worse the performance gain, but the decision is yours.